File: apps/api/src/vigia_api/domain/invites.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import StrEnum


class InviteStatus(StrEnum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    EXPIRED = "expired"
    REVOKED = "revoked"


@dataclass
class OrganizationInvite:
    id: str
    organization_id: str
    email: str
    role: str
    invited_by_user_id: str
    token_hash: str
    status: InviteStatus = InviteStatus.PENDING
    expires_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    accepted_at: datetime | None = None
    revoked_at: datetime | None = None
    accepted_by_user_id: str | None = None


@dataclass(frozen=True)
class QueuedEmail:
    id: str
    to_email: str
    subject: str
    body: str
    organization_id: str
    invite_id: str
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

File: apps/web/src/api/invites.ts
import { apiFetch, toJsonBody } from './client'

export type InviteStatus = 'pending' | 'accepted' | 'expired' | 'revoked'

export interface OrganizationInvite {
  id: string
  organization_id: string
  email: string
  role: string
  invited_by_user_id: string
  status: InviteStatus
  expires_at: string
  created_at: string
  updated_at: string
  accepted_at: string | null
  revoked_at: string | null
  accepted_by_user_id: string | null
}

export interface CreateInviteInput {
  email: string
  role: string
}

export function listInvites(organizationId: string) {
  return apiFetch<{ items: OrganizationInvite[] }>(`/organizations/${organizationId}/invites`)
}

export function createInvite(organizationId: string, payload: CreateInviteInput) {
  return apiFetch<{ invite: OrganizationInvite; token: string }>(`/organizations/${organizationId}/invites`, {
    method: 'POST',
    body: toJsonBody({ email: payload.email, role: payload.role }),
  })
}

export function resendInvite(organizationId: string, inviteId: string) {
  return apiFetch<{ token: string }>(`/organizations/${organizationId}/invites/${inviteId}:resend`, { method: 'POST' })
}

export function revokeInvite(organizationId: string, inviteId: string) {
  return apiFetch<{ invite: OrganizationInvite }>(`/organizations/${organizationId}/invites/${inviteId}:revoke`, { method: 'POST' })
}
